### PR TITLE
Introduce an option to allow nested lemma, and turn it off by default.

### PR DESCRIPTION
**Pull request by Zimmi48**
**Kind:** enhancement.

- [x] Corresponding documentation was added / updated.
- [x] Entry added in CHANGES.

The nested lemma feature has been bugging me for a while. It is a source of confusion for beginners who forget a `Qed`:

```
Lemma foo : True.
easy.
(* No more subgoals. *)

Lemma bar : True.
apply foo.
(* The reference foo was not found in the current environment. *)
```

It is also the reason why #7433 was not detected sooner.

Independently of these reasons, it has been deprecated for a while *but* we cannot actually remove the feature because lots of experimented users are actually relying on it.

So instead, what I propose in this PR is to create a new option to control whether the nested lemma feature is allowed or not. And turn this option off by default (this is the whole point). We could also (not necessarily, and not done currently in this PR) remove the deprecation warning... When we really intend to remove the feature, we can mark the option itself as deprecated (again, not done currently).

So what do people think of this proposal?

Implementation side, I was surprised that this was more difficult than I expected. I am still not sure I made the right choices so please review. I am not sure of the location of the option in particular (at the beginning it was next to the function raising the error message in `proof_global`).

### CODE REVIEW

**Comment by Zimmi48** [comment]
The code owner feature is definitely broken. Fortunately, we have @SkySkimmer tool in #7170 which gives:

```
vernac/lemmas.mli: @mattam82
vernac/lemmas.ml: @mattam82
proofs/pfedit.mli: @ppedrot
proofs/pfedit.ml: @ppedrot
proofs/proof_global.mli: @ppedrot
proofs/proof_global.ml: @ppedrot
```

**Comment by SkySkimmer** [comment]
Why make the argument optional?

**Comment by Zimmi48** [comment]
To do a lighter refactoring. But if you think it should be mandatory, I can adapt.

**Comment by SkySkimmer** [comment]
I certainly want to know which places if any need `allow_nested:true`.

**Comment by Zimmi48** [comment]
OK, I'll change this (not right now though).

**Comment by gares** [comment]
> So what do people think of this proposal?

thanks!

**Review by ejgallego — COMMENTED** [review]
I am not sure what to think of this:
- the good side is that it would remove lots of code in the current implementation,
- on the other hand, I think that supporting nested lemmas doesn't pose any particular technical problem, and the feature does indeed seem to make sense to quite a few experienced users. [The first particular technical problem is the duplication of functionalities among the STM and `Proof_global` which instead of working together fight against each other].

This last point brings us to a question on where to hook the warning, @Zimmi48 has chosen [_wisely IMHO_] to hook it in the execution layer, however this implies some particular tradeoffs, it could be indeed interesting to put the warning in the STM.

**Comment by Zimmi48** [comment]
@ejgallego It seems that the first part of your review comment is discussing whether to remove the nested lemma feature eventually. This PR does not take any side on this question.

**Comment by maximedenes** [comment]
> It seems that the first part of your review comment is discussing whether to remove the nested lemma feature eventually. This PR does not take any side on this question.

I think that's precisely the problem. I'm not in favor of introducing a flag if the plan is to make it stay. For example, this is going to make testing nested lemma more difficult.

That said, I'm in favor of removing the nested lemmas feature, after addressing the various user needs that they seem to meet today. The first step would probably be to provide a decent asynchronous IDE. On the other hand, I wouldn't cry if they stayed, if properly supported.

So in short, if we agree to remove nested lemmas in the long term, I'm in favor of this PR, otherwise I'm against.

**Comment by Zimmi48** [comment]
> Because you need to test interactions between features. Adding flags is an obvious source of exponential growth in the number of test cases needed.

Oh but here the flag does not create two widely different behaviors that would need to be tested against each orthogonal feature, it's more like a warning. So I don't think that this (otherwise valid) argument applies. In particular, having the feature on by default doesn't mean that we properly test it.

**Comment by maximedenes** [comment]
> the flag does not create two widely different behaviors that would need to be tested against each orthogonal feature

Ok, maybe you can argue that this is a special case. The `Universe Polymorphic` flag you took as an example definitely does not fall in this category, though.

**Comment by Zimmi48** [comment]
Indeed, that was a pretty bad comparison.

**Comment by ejgallego** [comment]
Maybe I am confused as to why do you want to turn this option on by default. This IMHO looks as a "Coq 9" breakage-level, quite a few users do have nested stuff.

**Comment by ejgallego** [comment]
In that sense my point of view is that I tend to consider the current warning the one that should go, but YMMV.

**Comment by ejgallego** [comment]
I see @Zimmi48 , umm, IMVVHO I think that the plan for the future of proof scripts themselves is not very explicitly stated yet, so I am not sure that doing this change without discussing a bit more about this is the best action path.

IMHO I would just wait for the two upcomign WG/CIM and see where are we going on the proof script sense.

**Comment by Zimmi48** [comment]
> so I am not sure that doing this change without discussing a bit more about this is the best action path.

Again, the goal of this PR was not to take a side on the debate of whether nested lemmas should go or not. What is the problem of considering this an expert-only feature and hiding it behind an option? It's not much different from having to do `Require Extraction` to use extraction. The error message even provides the fix.

**Comment by ejgallego** [comment]
IMHO setting the option OFF by default looks a bit like gratuitous breakage, specially if after discussion we decide to keep nested lemmas in. To me, the default only makes sense if we are set on having nested lemmas go, and IMHO we should discuss that a bit more if possible.

**Comment by Zimmi48** [comment]
It's not gratuitous. The goal is to solve a real issue I have been confronted to several times which is that when people forget a `Qed` this feature prevents them from having any warning or error message. That being said, the balance is tipping against this change so I'm not going to work more on this PR unless I get real support (for now, the only support I got was from @gares).

**Comment by Zimmi48** [comment]
Having this option on by default would (AFAICS) have zero utility.

**Comment by SkySkimmer** [comment]
I very approve of disabling nesting by default.

**Comment by ejgallego** [comment]
> I'm also supporting the fact that nested lemmas should be disabled by default.

What are the problems @ppedrot you see with nested lemmas [honest question, I'm curious]

I'd dare to say that in maths nested / auxiliary lemmas are quite common, thus hence my reluctance.

**Comment by maximedenes** [comment]
> I'd dare to say that in maths nested / auxiliary lemmas are quite common, thus hence my reluctance.

Auxiliary, obviously. But nested, do you have an example?

**Comment by Zimmi48** [comment]
> I am not "against" this change per se, but I'd like to avoid the situation where we push the change, then we declare that nested lemmas are OK.

Once again, it looks like I failed to get that clear: this PR is not taking any side on whether nested lemmas are a good or a bad feature. It is just classifying it as an "advanced" feature that is fine to disable by default (again, like Extraction for instance).

**Comment by Zimmi48** [comment]
But I'm OK with alternative syntax proposal if you prefer this. This would indeed solve my problem.

**Comment by ejgallego** [comment]
> Auxiliary, obviously. But nested, do you have an example?

I fail to locate from memory any paper that is not written by me and states a lemma inside a proof, but I'd dare to say there are quite a few.

**Comment by ppedrot** [comment]
> What are the problems @ppedrot you see with nested lemmas [honest question, I'm curious]

I've also been bitten by the same kind of problems that is mentioned in the first post, i.e. an unclosed proof. Although a trivial problem, I find it rather frustrating that Coq cannot tell me I forgot something.

> I'd dare to say that in maths nested / auxiliary lemmas are quite common, thus hence my reluctance.

Nested lemmas are actually not nested, because they're accessible from the outside. In Coq they're just a syntactic artifact, they don't bring anything at deployment time. While writing a proof it's another story, but then I believe it's mostly an UI issue.

**Comment by ejgallego** [comment]
> It is just classifying it as an "advanced" feature that is fine to disable by default

Yeah in a sense I am questioning if making the vernacular conditional on an "advanced" flag is the way to go, here indeed we'd face indeed option creep. I'd say that I'd prefer a vernacular with few options like those. So we either fully embrace such style of proof, or we get rid of it.

> (again, like Extraction for instance).

IMHO the comparison is not fully fair, in the sense that extraction was removed from the init file as to make startup lighter, but this change doesn't do anything like that.

**Comment by ejgallego** [comment]
So indeed the main problem to solve here seems to be bad parenthezization.

> they don't bring anything at deployment time.

The only benefit is for authors to organize their development like that.

> While writing a proof it's another story, but then I believe it's mostly an UI issue.

Well, in a sense the "I forgot to close a proof" issue seems indeed very close to be an UI issue too.

By the way, I think that true "nested" lemmas are pervasive in developments such as mathcomp, it is just that they use `have` of `suff` instead of the Coq machinery.

**Comment by ejgallego** [comment]
> By the way, I think that true "nested" lemmas are pervasive in developments such as mathcomp, it is just that they use have of suff instead of the Coq machinery.

So let me try to rephrase my worry a bit more once that I have mentioned this, and independently of this proposal that maybe is the right thing to do.

"How would we like the document vernacular (*) to reflect the internal proof structure?"

Indeed, it would make sense to me, that the vernacular would provide structure for subproofs `have`, `suff`, and `case` analysis, side obligations at least. But indeed the current sub lemma support is not bringing anything very useful here, IMHO.

So for me parenthesization is a valid reason to disable "nesting" of lemmas, however the argument that such pattern cannot be properly supported (used in the past) is not a valid one AFAIAC.

(*) Document vernacular here == structure the UI understands.

**Comment by Zimmi48** [comment]
> By the way, I think that true "nested" lemmas are pervasive in developments such as mathcomp, it is just that they use have of suff instead of the Coq machinery.

That's not the same at all as these depend on the local context. And vanilla Coq supports this too with `assert` and `enough`.

> So for me parenthesization is a valid reason to disable "nesting" of lemmas, however the argument that such pattern cannot be properly supported (used in the past) is not a valid one AFAIAC.

Thanks. Indeed, I wasn't using this other argument.

**Comment by ejgallego** [comment]
> Thanks. Indeed, I wasn't using this other argument.

Yeah so indeed IIUC the deprecation was done for that other reasons. I guess the current proposal looks OK to me then.

**Comment by Zimmi48** [comment]
I made the `~allow_nested` parameter mandatory as recommended by @SkySkimmer and I added a commit that removes the deprecation warning. Documentation will wait because I got distracted by the state of the manual.

File: proofs/pfedit.ml
Comment (Zimmi48) [inline proofs/pfedit.ml]: `build_constant_by_tactic` is used by `abstract`.

File: vernac/comFixpoint.ml
Comment (Zimmi48) [inline vernac/comFixpoint.ml]: I haven't fully understood what `declare_fixpoint` does but it may well be that this should be set to `false` instead.

File: vernac/lemmas.ml
Comment (Zimmi48) [inline vernac/lemmas.ml]: `start_proof_com` is the one used by `Lemma` and the like.

File: vernac/obligations.ml
Comment (Zimmi48) [inline vernac/obligations.ml]: I don't know really what this code does.

File: vernac/lemmas.ml
Comment (Zimmi48) [inline vernac/lemmas.ml]: I could also add the `~allow_nested` parameter to `Lemmas.start_proof` to show further where it's being called (it is called in Funind in particular).

File: vernac/comFixpoint.ml
Comment (SkySkimmer) [reply vernac/comFixpoint.ml]: I think this is what happens for `Fixpoint foo (n:nat) : True.` so it should allow nested iff the option is on.

File: vernac/lemmas.ml
Comment (SkySkimmer) [reply vernac/lemmas.ml]: For instance the call in classes.ml should respect the option.

Comment (Zimmi48) [reply vernac/lemmas.ml]: OK

Comment (ejgallego) [reply vernac/lemmas.ml]: Yup there are quite a few callers, so maybe you want instead of plug the option at the STM level? In fact `pstates` may be close to be gone, you can check nesting quite easily in `process_transaction` I think.

**Review by Zimmi48 — COMMENTED** [review]
Thanks @ejgallego. I followed your suggestion of putting the check for nested lemma in the STM. However, I need help on how to raise an error here.

The previous version has been safely saved at https://github.com/Zimmi48/coq/tree/nested-lemma/first-trial

File: stm/stm.ml
Comment (ejgallego) [reply stm/stm.ml]: I'm much afraid that there is no "support" for your use case, that anomaly seems in general bogus as many things can go wrong without a `safe_id`. What the code does now is to use `Exninfo.iraise (State.exn_on ~valid Stateid.dummy exn)`, that may be the most sensible thing to do given the current code.

File: stm/stm.ml
Comment (ejgallego) [inline stm/stm.ml]: While you are at it make this explicit `    | `Sideff (CherryPickEnv,_) -> `

**Comment by Zimmi48** [comment]
@ejgallego I tried to implement your suggestion but something is off because the fake-ide tests which involve nested lemmas (and to which I added the option) are failing (they were not in the initial version).

**Comment by ejgallego** [comment]
> @ejgallego I tried to implement your suggestion but something is off because the fake-ide tests which involve nested lemmas (and to which I added the option) are failing (they were not in the initial version).

I'd be glad to have a look if you think that could be of help [next week tho]

**Comment by gares** [comment]
I think `sed -i 'somethink'` works on both osx (non-gnu) sed and linux sed, so the changes are good to me.

**Comment by gares** [comment]
I'm looking at the failures right now

**Comment by gares** [comment]
The reason is subtle. Since the value of the option affects the *scheduling* phase of the STM (rather than just the *execution* of commands), the value of the bool ref has to be update before scheduling takes place.
The same happens with `Set Universe Polymorphism`.

I'm pushing to your branch the same fix I used for the other flag: the classifier tells the STM that such command has to be executed immediately (eg as "Notation").

The nicer fix would be to make the STM accumulate these options while doing the scheduling, or having the attribute system take care of that (IIRC I discussed it with @maximedenes and @vbgl ). The fact (attribute) that a Lemma admits nested (sub) lemmas would need the same treatment of `Set Universe Polymorphism` and the "being polymorphic" attribute. Hopefully there two quirks will go away together.

**Review by gares — APPROVED** [review]

**Comment by Zimmi48** [comment]
Thanks a lot @gares! I'm waiting on the merge of #7468 and #7461 before adding the doc to this PR.

**Comment by Zimmi48** [comment]
The overlays are done and are incredibly simple: only one file of Color and one file of Corn required the option. This confirms that people don't use this feature much in finished code even if they rely on it during development.

The documentation remains to be done: I got side tracked one again (cf. #7487).

File: stm/vernac_classifier.ml
Comment (ejgallego) [inline stm/vernac_classifier.ml]: Spelling of `scheuling`

**Comment by ejgallego** [comment]
Indeed you can remove the overlays now I think. I support this patch, will review once the doc is in place.

It also needs an assignee.

**Comment by Zimmi48** [comment]
This PR is now ready. The overlays have been dropped and the documentation has been added. Did you know that there wasn't a proper documentation of the nested lemma feature so far?
Regarding the assignee, given that the only affected component is the STM, it is either @gares or @ejgallego.

File: CHANGES
Comment (ejgallego) [inline CHANGES]: Looks fine but maybe I would tweak the phrasing a bit more. "Using "nested proofs", that is to say, opening a `Lemma` inside the proof of another `Lemma` will produce now an error by default. You can keep using them by activating the `Nested Proof Allowed` option."

**Review by ejgallego — APPROVED** [review]
LGTM.

File: CHANGES
Comment (Zimmi48) [reply CHANGES]: Thanks for your comment. I'm unconvinced by my current formulation, but I'm not sure I really prefer the one you are proposing. Any other suggestion anyone?

Comment (SkySkimmer) [reply CHANGES]: Nested proofs may be enabled through the option `...`. By default they are disabled and produce an error. The deprecation warning which used to occur when using nested proofs has been removed.

Comment (Zimmi48) [reply CHANGES]: Thanks @SkySkimmer!

**Comment by Zimmi48** [comment]
This now depends on #7517.

File: CHANGES
Comment (ejgallego) [reply CHANGES]: I think I prefer a phrasing that says first "now nested proofs will produce an error"

This is the most relevant bit of information for the user "action X, that used to work, now produces an error." Then, we should add some info how to fix it.

**Comment by ejgallego** [comment]
> Regarding the assignee, given that the only affected component is the STM, it is either @gares or @ejgallego.

I am not sure how things are supposed to work here, as in principle I'd have to be delegated by the main maintainer to take over this PR [which I'd be happy to do].

On the other hand Enrico has a small patch here, so I dunno. Feel free to assign me if you want me to drive the merging of this PR.

**Comment by Zimmi48** [comment]
Yeah, I feel like @gares kind of co-authored this PR so it would make sense that you drive it.

**Review by ejgallego — CHANGES_REQUESTED** [review]
Our friend the linter complains.

**Review by ejgallego — APPROVED** [review]
Latest nits are addressed, this is going to be merged very soon, last call to complain!

**Comment by ejgallego** [comment]
Oh I just see this needs a rebase, sorry @Zimmi48 !

**Comment by Zimmi48** [comment]
No worries, that was a trivial one.

**Pull request merged**
